File: src/cli/registry.ts
import { Command } from "../commands/types.js";
import { AddCommand } from "../commands/AddCommand.js";
import { BranchCommand } from "../commands/BranchCommand.js";
import { CatFileCommand } from "../commands/CatFileCommand.js";
import { CommitCommand } from "../commands/CommitCommand.js";
import { DiffCommand } from "../commands/DiffCommand.js";
import { HashObjectCommand } from "../commands/HashObjectCommand.js";
import { InitCommand } from "../commands/InitCommand.js";
import { LogCommand } from "../commands/LogCommand.js";
import { LSTreeCommand } from "../commands/LSTreeCommand.js";
import { RestoreCommand } from "../commands/RestoreCommand.js";
import { StatusCommand } from "../commands/StatusCommand.js";
import { WriteTreeCommand } from "../commands/WriteTreeCommand.js";
import { InvalidCommandError } from "../common/helpers/errors/minigit.js";

const registry = new Map<string, Command>([
  ["add", new AddCommand()],
  ["branch", new BranchCommand()],
  ["cat-file", new CatFileCommand()],
  ["commit", new CommitCommand()],
  ["diff", new DiffCommand()],
  ["hash-object", new HashObjectCommand()],
  ["init", new InitCommand()],
  ["log", new LogCommand()],
  ["ls-tree", new LSTreeCommand()],
  ["restore", new RestoreCommand()],
  ["status", new StatusCommand()],
  ["write-tree", new WriteTreeCommand()],
]);

export const getCommand = (name: string): Command => {
  const command = registry.get(name);
  if (command) return command;

  throw new InvalidCommandError(`unknown command: ${name}`);
};

File: src/cli/registry.cpp
#include "minigit/cli/registry.hpp"
#include "minigit/cli/commands/add_command.hpp"
#include "minigit/cli/commands/branch_command.hpp"
#include "minigit/cli/commands/cat_file_command.hpp"
#include "minigit/cli/commands/commit_command.hpp"
#include "minigit/cli/commands/diff_command.hpp"
#include "minigit/cli/commands/hash_object_command.hpp"
#include "minigit/cli/commands/init_command.hpp"
#include "minigit/cli/commands/log_command.hpp"
#include "minigit/cli/commands/ls_tree_command.hpp"
#include "minigit/cli/commands/restore_command.hpp"
#include "minigit/cli/commands/status_command.hpp"
#include "minigit/cli/commands/write_tree_command.hpp"
#include "minigit/common/util/error.hpp"

namespace minigit::cli {

CommandRegistry::CommandRegistry()
{
    register_command("init", std::make_unique<InitCommand>());
    register_command("hash-object", std::make_unique<HashObjectCommand>());
    register_command("cat-file", std::make_unique<CatFileCommand>());
    register_command("write-tree", std::make_unique<WriteTreeCommand>());
    register_command("ls-tree", std::make_unique<LsTreeCommand>());
    register_command("add", std::make_unique<AddCommand>());
    register_command("restore", std::make_unique<RestoreCommand>());
    register_command("status", std::make_unique<StatusCommand>());
    register_command("commit", std::make_unique<CommitCommand>());
    register_command("log", std::make_unique<LogCommand>());
    register_command("branch", std::make_unique<BranchCommand>());
    register_command("diff", std::make_unique<DiffCommand>());
}

void CommandRegistry::register_command(std::string_view name, std::unique_ptr<Command> command)
{
    commands_.emplace(std::string(name), std::move(command));
}

const Command& CommandRegistry::get(std::string_view name) const
{
    const auto it = commands_.find(std::string(name));
    if (it == commands_.end())
    {
        throw Error(ErrorCode::InvalidCommand,
                    "INVALID_COMMAND_ERROR: unknown command: " + std::string(name));
    }
    return *it->second;
}

} // namespace minigit::cli
